core/src/banking_trace.rs

```diff
@@ -0,0 +1,549 @@
+            is_checked: bool::default(),
+        }
+    }
+
+    fn reset(&mut self) {
+        self.is_checked = false;
+    }
+}
+
+struct GroupedWriter<'a> {
+    now: DateTime<Local>,
+    underlying: &'a mut RollingFileAppender<RollingConditionGrouped>,
+}
+
+impl<'a> GroupedWriter<'a> {
+    fn new(underlying: &'a mut RollingFileAppender<RollingConditionGrouped>) -> Self {
+        Self {
+            now: Local::now(),
+            underlying,
+        }
+    }
+}
+
+impl RollingCondition for RollingConditionGrouped {
+    fn should_rollover(&mut self, now: &DateTime<Local>, current_filesize: u64) -> bool {
+        if !self.tried_rollover_after_opened {
+            self.tried_rollover_after_opened = true;
+
+            // rollover normally if empty to reuse it if possible
+            if current_filesize > 0 {
+                // forcibly rollover anew, so that we always avoid to append
+                // to a possibly-damaged tracing file even after unclean
+                // restarts
+                return true;
+            }
+        }
+
+        if !self.is_checked {
+            self.is_checked = true;
+            self.basic.should_rollover(now, current_filesize)
+        } else {
+            false
+        }
+    }
+}
+
+impl<'a> Write for GroupedWriter<'a> {
+    fn write(&mut self, buf: &[u8]) -> std::result::Result<usize, io::Error> {
+        self.underlying.write_with_datetime(buf, &self.now)
+    }
+    fn flush(&mut self) -> std::result::Result<(), io::Error> {
+        self.underlying.flush()
+    }
+}
+
+pub fn receiving_loop_with_minimized_sender_overhead<T, E, const SLEEP_MS: u64>(
+    exit: Arc<AtomicBool>,
+    receiver: Receiver<T>,
+    mut on_recv: impl FnMut(T) -> Result<(), E>,
+) -> Result<(), E> {
+    'outer: while !exit.load(Ordering::Relaxed) {
+        'inner: loop {
+            // avoid futex-based blocking here, otherwise a sender would have to
+            // wake me up at a syscall cost...
+            match receiver.try_recv() {
+                Ok(message) => on_recv(message)?,
+                Err(TryRecvError::Empty) => break 'inner,
+                Err(TryRecvError::Disconnected) => {
+                    break 'outer;
+                }
+            };
+            if exit.load(Ordering::Relaxed) {
+                break 'outer;
+            }
+        }
+        sleep(Duration::from_millis(SLEEP_MS));
+    }
+
+    Ok(())
+}
+
+impl BankingTracer {
+    pub fn new(
+        maybe_config: Option<(PathBuf, Arc<AtomicBool>, DirByteLimit)>,
+    ) -> Result<Arc<Self>, TraceError> {
+        let enabled_tracer = maybe_config
+            .map(|(path, exit, dir_byte_limit)| {
+                let rotate_threshold_size = dir_byte_limit / TRACE_FILE_ROTATE_COUNT;
+                if rotate_threshold_size == 0 {
+                    return Err(TraceError::TooSmallDirByteLimit(
+                        dir_byte_limit,
+                        TRACE_FILE_ROTATE_COUNT,
```

Comment:
Do you know how wrapping works in the rolling file? If I have 10 bytes remaining in my desired size, and try to write 20 bytes will it roll the the next file, or just "overflow" the file by 10 bytes?

Just curious if there's an edge case here - if the user specifies 14 bytes as the limit, this check passes and then each file has a threshold size of 1 byte. That'd be too small to write any of the events, so I'm hoping that it would just overflow.

Reply:
yep, it just overflow: a84bc5501d

Reply:
awesome, that's what I was hoping for